File: Data Structures & Algorithms/binary-tree-diameter/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter: int = 0

        def find_diameter(root) -> int:
            nonlocal diameter
            if not root:
                return 0

            left_diameter = find_diameter(root.left)
            right_diameter = find_diameter(root.right)
            subtree_diameter = left_diameter + right_diameter
            diameter = max(diameter, subtree_diameter)
            return 1 + max(left_diameter, right_diameter)

        find_diameter(root)
        return diameter

File: Data Structures & Algorithms/valid-palindrome-ii/submission-6.java
class Solution {
    public boolean isPalindrome(String s, int left, int right){
        while(left < right){
            if(s.charAt(right) != s.charAt(left)){
                return false;
            }
            ++left;
            --right;
        }
        return true;
    }
    public boolean validPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;

        while( i < j){
            if(s.charAt(i) != s.charAt(j)){
                return isPalindrome(s, i, j-1) || isPalindrome(s, i+1, j);
            } 
            ++i;
            --j;
        }
        return true;
    }
    
}